lib/cylc/ws_data_mgr.py

```diff
@@ -0,0 +1,394 @@
+                    setattr(family.meta, key, val)
+                else:
+                    family.meta.user_defined.append(f"{key}={val}")
+            family.parents.extend(
+                [f"{self.workflow_id}/{p_name}"
+                    for p_name in parents_dict[name]])
+            families[name] = family
+
+        for name, parent_list in parents_dict.items():
+            if parent_list and parent_list[0] in families:
+                if name in config.taskdefs:
+                    families[parent_list[0]].child_tasks.append(name)
+                else:
+                    families[parent_list[0]].child_families.append(name)
+
+        for point_string, c_task_states in self.task_states.items():
+            # For each cycle point, construct a family state tree
+            # based on the first-parent single-inheritance tree
+
+            c_fam_task_states = {}
+
+            count = {}
+
+            for key in c_task_states:
+                state = c_task_states[key]
+                if state is None:
+                    continue
+                try:
+                    count[state] += 1
+                except KeyError:
+                    count[state] = 1
+
+                all_states.append(state)
+                for parent in ancestors_dict.get(key, []):
+                    if parent == key:
+                        continue
+                    c_fam_task_states.setdefault(parent, set([]))
+                    c_fam_task_states[parent].add(state)
+
+            state_count_cycles[point_string] = count
+
+            for fam, child_states in c_fam_task_states.items():
+                state = extract_group_state(child_states)
+                if state is None or fam not in families:
+                    continue
+                int_id = TaskID.get(fam, point_string)
+                fp_id = f"{self.workflow_id}/{point_string}/{fam}"
+                fp_check = f"{int_id}@{update_time}"
+                families[fam].proxies.append(fp_id)
+                fproxy = PbFamilyProxy(
+                    checksum=fp_check,
+                    id=fp_id,
+                    cycle_point=point_string,
+                    name=fam,
+                    family=f"{self.workflow_id}/{fam}",
+                    state=state,
+                    depth=len(ancestors_dict[fam]) - 1,
+                )
+                for child_name in descendants_dict[fam]:
+                    ch_id = f"{self.workflow_id}/{point_string}/{child_name}"
+                    if parents_dict[child_name][0] == fam:
+                        if child_name in c_fam_task_states:
+                            fproxy.child_families.append(ch_id)
+                        else:
+                            fproxy.child_tasks.append(ch_id)
+                fproxy.parents.extend(
+                    [f"{self.workflow_id}/{point_string}/{p_name}"
+                        for p_name in parents_dict[name]])
```

Comment:
Are you intentionally using `name` variable here that was created earlier in another `for` loop? (`for name, parent_list in parents_dict.items():`, line 181).

Reply:
Yes - The `name` variable there refers to the family whose parents we want to find.